syz-cluster/controller: TestProcessor times out

This will likely be gone after the refactoring I'm going to push for review soon, but let's keep it in the issues for now.

https://github.com/google/syzkaller/actions/runs/13296004987/job/37128005699
```
2025-02-12T22:35:19.1020779Z panic: test timed out after 10m0s
2025-02-12T22:35:19.1021197Z 	running tests:
2025-02-12T22:35:19.1021543Z 		TestProcessor (9m59s)
2025-02-12T22:35:19.1021778Z 
2025-02-12T22:35:19.1021939Z goroutine 6743 [running]:
2025-02-12T22:35:19.1022336Z testing.(*M).startAlarm.func1()
2025-02-12T22:35:19.1023133Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/testing/testing.go:2373 +0x385
2025-02-12T22:35:19.1023954Z created by time.goFunc
2025-02-12T22:35:19.1024656Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/time/sleep.go:215 +0x2d
2025-02-12T22:35:19.1025307Z 
< ... >
2025-02-12T22:35:19.1035592Z main.main()
2025-02-12T22:35:19.1036027Z 	_testmain.go:63 +0x9b
2025-02-12T22:35:19.1036294Z 
2025-02-12T22:35:19.1036527Z goroutine 6 [select]:
2025-02-12T22:35:19.1037232Z go.opencensus.io/stats/view.(*worker).start(0xc000356580)
2025-02-12T22:35:19.1038156Z 	/syzkaller/.cache/gomod/go.opencensus.io@v0.24.0/stats/view/worker.go:292 +0x9f
2025-02-12T22:35:19.1039084Z created by go.opencensus.io/stats/view.init.0 in goroutine 1
2025-02-12T22:35:19.1039917Z 	/syzkaller/.cache/gomod/go.opencensus.io@v0.24.0/stats/view/worker.go:34 +0x8d
2025-02-12T22:35:19.1040470Z 
2025-02-12T22:35:19.1040631Z goroutine 517 [chan send, 9 minutes]:
2025-02-12T22:35:19.1041566Z github.com/google/syzkaller/syz-cluster/controller.TestProcessor(0xc000bf2b60)
2025-02-12T22:35:19.1042756Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor_test.go:73 +0x647
2025-02-12T22:35:19.1043708Z testing.tRunner(0xc000bf2b60, 0x23e2268)
2025-02-12T22:35:19.1044541Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/testing/testing.go:1690 +0xf4
2025-02-12T22:35:19.1045384Z created by testing.(*T).Run in goroutine 1
2025-02-12T22:35:19.1046256Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/testing/testing.go:1743 +0x390
```

No, it didn't go away: https://github.com/google/syzkaller/actions/runs/13396267430/job/37415981700?pr=5800

```
2025-02-18T17:27:17.1756018Z panic: test timed out after 10m0s
2025-02-18T17:27:17.1756481Z 	running tests:
2025-02-18T17:27:17.1756820Z 		TestProcessor (10m0s)
2025-02-18T17:27:17.1757071Z 
< ... >

2025-02-18T17:27:17.1777910Z goroutine 24 [chan send, 9 minutes]:
2025-02-18T17:27:17.1778684Z github.com/google/syzkaller/syz-cluster/controller.TestProcessor(0xc00019b520)
2025-02-18T17:27:17.1779857Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor_test.go:74 +0x618
2025-02-18T17:27:17.1780803Z testing.tRunner(0xc00019b520, 0x2506058)
2025-02-18T17:27:17.1781598Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/testing/testing.go:1690 +0xf4
2025-02-18T17:27:17.1782444Z created by testing.(*T).Run in goroutine 1
2025-02-18T17:27:17.1783291Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/testing/testing.go:1743 +0x390
2025-02-18T17:27:17.1784345Z 
< .. >

2025-02-18T17:27:17.2128068Z goroutine 311 [semacquire, 9 minutes]:
2025-02-18T17:27:17.2128510Z sync.runtime_Semacquire(0x447c9f?)
2025-02-18T17:27:17.2129271Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/runtime/sema.go:71 +0x25
2025-02-18T17:27:17.2130069Z sync.(*WaitGroup).Wait(0x1da17d5?)
2025-02-18T17:27:17.2130839Z 	/syzkaller/.cache/gomod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/sync/waitgroup.go:118 +0x48
2025-02-18T17:27:17.2132083Z github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).Loop(0xc000982780, {0x2738038, 0xc00095adc0})
2025-02-18T17:27:17.2133703Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:81 +0x397
2025-02-18T17:27:17.2135022Z github.com/google/syzkaller/syz-cluster/controller.TestProcessor.func2()
2025-02-18T17:27:17.2136145Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor_test.go:63 +0x28
2025-02-18T17:27:17.2137356Z created by github.com/google/syzkaller/syz-cluster/controller.TestProcessor in goroutine 24
2025-02-18T17:27:17.2138712Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor_test.go:62 +0x59b
< ... >


2025-02-18T17:27:17.2294695Z goroutine 121 [select]:
2025-02-18T17:27:17.2295640Z github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).streamSeries(0xc000982780, {0x2738038, 0xc00095adc0}, 0xc000ac93b0)
2025-02-18T17:27:17.2297115Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:87 +0xe5
2025-02-18T17:27:17.2298491Z github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).Loop.func2()
2025-02-18T17:27:17.2299616Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:78 +0x66
2025-02-18T17:27:17.2301045Z created by github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).Loop in goroutine 311
2025-02-18T17:27:17.2302504Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:76 +0x375

< .. >


2025-02-18T17:27:17.5168876Z make[1]: *** [Makefile:394: test] Error 1
2025-02-18T17:27:17.5172290Z make: *** [Makefile:323: presubmit_build] Error 2
2025-02-18T17:27:17.5177228Z github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).seriesRunner(0xc000982780, {0x2738038, 0xc00095adc0}, 0xc000ac93b0)
2025-02-18T17:27:17.5179737Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:116 +0x167
2025-02-18T17:27:17.5181107Z github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).Loop.func1()
2025-02-18T17:27:17.5182243Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:63 +0x5d
2025-02-18T17:27:17.5183615Z created by github.com/google/syzkaller/syz-cluster/controller.(*SeriesProcessor).Loop in goroutine 311
2025-02-18T17:27:17.5184906Z 	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/syz-cluster/controller/processor.go:61 +0x15d
2025-02-18T17:27:17.5185949Z FAIL	github.com/google/syzkaller/syz-cluster/controller	600.075s
```

--- a/syz-cluster/controller/processor.go
+++ b/syz-cluster/controller/processor.go
@@ -154,11 +154,16 @@ func (sp *SeriesProcessor) handleSession(ctx context.Context, session *db.Sessio
 		switch status {
 		case workflow.StatusNotFound:
 			log.Printf("scheduling a workflow for %q", session.ID)
-			if err := sp.sessionRepo.Start(ctx, session.ID); err != nil {
+			err := sp.sessionRepo.Start(ctx, session.ID)
+			if err == db.ErrSessionAlreadyStarted {
+				// It may happen if the service was restarted right between the moment we updated the DB
+				// and actually started the workflow.
+				log.Printf("session %q was already marked as started, but there's no actual workflow", session.ID)
+			} else if err != nil {
 				app.Errorf("failed to mark session started: %v", err)
 				break
 			}
-			err := sp.workflows.Start(session.ID)
+			err = sp.workflows.Start(session.ID)
 			if err != nil {
 				app.Errorf("failed to start a workflow: %v", err)
 			}

--- a/syz-cluster/controller/processor_test.go
+++ b/syz-cluster/controller/processor_test.go
@@ -52,7 +52,9 @@ func TestProcessor(t *testing.T) {
 
 	awaitFinishedSessions(t, processor.seriesRepo, 2)
 
-	// Restart the loop.
+	// Emulate the service restart by aborting the loop.
+	// This may break the execution in arbitrary places, which actually resembles the environment in which the code
+	// will actually work. The bugs it triggers may be difficult to reproduce though.
 	cancel()
 	wg.Wait()
 

